File: src/tui/theme/store.ts
import type { ResolvedTheme, ThemeJson, ThemeMode, ThemeState, ThemeStore } from "./types";
import { resolveTheme } from "./resolver";
import { DEFAULT_THEME_ID } from "./builtin";
import { ThemeRegistry } from "./registry";

type Listener = () => void;

export interface PersistedThemeState {
  active: string;
  mode?: ThemeMode;
  lock?: ThemeMode;
}

export class FrostThemeStore implements ThemeStore {
  private registry: ThemeRegistry;
  private state: ThemeState;
  private listeners: Set<Listener> = new Set();
  private resolvedCache: Map<string, ResolvedTheme> = new Map();
  private persistCallback: ((state: PersistedThemeState) => void) | null = null;

  constructor(registry: ThemeRegistry) {
    this.registry = registry;
    this.state = {
      themes: registry.getAll(),
      mode: "dark",
      lock: undefined,
      active: DEFAULT_THEME_ID,
      ready: false,
    };
  }

  setPersistCallback(cb: (state: PersistedThemeState) => void): void {
    this.persistCallback = cb;
  }

  private notify(): void {
    this.state.themes = this.registry.getAll();
    for (const cb of this.listeners) {
      try {
        cb();
      } catch {
        // ignore listener errors
      }
    }
  }

  private persist(): void {
    if (this.persistCallback) {
      this.persistCallback({
        active: this.state.active,
        mode: this.state.lock ? this.state.mode : undefined,
        lock: this.state.lock,
      });
    }
  }

  private invalidateCache(): void {
    this.resolvedCache.clear();
  }

  getState(): ThemeState {
    return { ...this.state, themes: this.registry.getAll() };
  }

  subscribe(cb: Listener): () => void {
    this.listeners.add(cb);
    return () => {
      this.listeners.delete(cb);
    };
  }

  getActive(): string {
    return this.state.active;
  }

  has(id: string): boolean {
    if (id === "system") return this.registry.getSystemTheme() !== null;
    return this.registry.has(id);
  }

  set(id: string): void {
    if (!this.registry.has(id)) return;
    if (this.state.active === id) return;
    this.state.active = id;
    this.invalidateCache();
    this.notify();
    this.persist();
  }

  getMode(): ThemeMode {
    return this.state.mode;
  }

  setMode(mode: ThemeMode): void {
    if (this.state.mode === mode) return;
    this.state.mode = mode;
    this.invalidateCache();
    this.notify();
    if (this.state.lock) {
      this.persist();
    }
  }

  getLock(): ThemeMode | undefined {
    return this.state.lock;
  }

  lock(mode: ThemeMode): void {
    this.state.lock = mode;
    this.state.mode = mode;
    this.invalidateCache();
    this.notify();
    this.persist();
  }

  unlock(): void {
    if (!this.state.lock) return;
    this.state.lock = undefined;
    this.notify();
    this.persist();
  }

  isReady(): boolean {
    return this.state.ready;
  }

  setReady(ready: boolean): void {
    this.state.ready = ready;
    this.notify();
  }

  getAll(): Record<string, ThemeJson> {
    return this.registry.getAll();
  }

  getTheme(id: string): ThemeJson | undefined {
    return this.registry.get(id);
  }

  getActiveKeys(): string[] {
    return this.registry.getIds();
  }

  resolveActive(): ResolvedTheme | null {
    const theme = this.registry.get(this.state.active);
    if (!theme) return null;

    const cacheKey = `${this.state.active}:${this.state.mode}`;
    const cached = this.resolvedCache.get(cacheKey);
    if (cached) return cached;

    try {
      const resolved = resolveTheme(theme, this.state.mode);
      this.resolvedCache.set(cacheKey, resolved);
      return resolved;
    } catch {
      return null;
    }
  }

  // Registry delegation methods

  merge(themes: Record<string, ThemeJson>): void {
    this.registry.merge(themes);
    this.invalidateCache();
    this.notify();
  }

  upsert(id: string, theme: ThemeJson): void {
    this.registry.upsert(id, theme);
    this.invalidateCache();
    this.notify();
  }

  remove(id: string): boolean {
    const result = this.registry.remove(id);
    if (result) {
      if (this.state.active === id) {
        this.state.active = this.registry.has(DEFAULT_THEME_ID)
          ? DEFAULT_THEME_ID
          : this.registry.getIds()[0] ?? DEFAULT_THEME_ID;
      }
      this.invalidateCache();
      this.notify();
    }
    return result;
  }

  setSystemTheme(theme: ThemeJson | null, currentActive?: string): void {
    const wasSystem = this.state.active === "system";
    this.registry.setSystemTheme(theme);
    this.invalidateCache();

    if (!theme && wasSystem) {
      const fallback = currentActive ?? DEFAULT_THEME_ID;
      if (this.registry.has(fallback)) {
        this.state.active = fallback;
      } else {
        this.state.active = DEFAULT_THEME_ID;
      }
    }

    this.notify();
  }

  getRegistry(): ThemeRegistry {
    return this.registry;
  }
}

File: crates/frost-core/src/theme/store.rs
use crate::theme::builtin::DEFAULT_THEME_ID;
use crate::theme::registry::ThemeRegistry;
use crate::theme::resolver::resolve_theme_safe;
use crate::theme::types::{ResolvedTheme, ThemeJson, ThemeMode};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use tokio::sync::broadcast;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PersistedThemeState {
    pub active: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mode: Option<ThemeMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lock: Option<ThemeMode>,
}

pub struct ThemeStore {
    registry: ThemeRegistry,
    active: String,
    mode: ThemeMode,
    lock: Option<ThemeMode>,
    resolved_cache: HashMap<(String, ThemeMode), ResolvedTheme>,
    persist_path: Option<PathBuf>,
    notify_tx: broadcast::Sender<()>,
}

impl ThemeStore {
    pub fn new(registry: ThemeRegistry) -> Self {
        let (notify_tx, _) = broadcast::channel(16);
        Self {
            registry,
            active: DEFAULT_THEME_ID.to_string(),
            mode: ThemeMode::Dark,
            lock: None,
            resolved_cache: HashMap::new(),
            persist_path: None,
            notify_tx,
        }
    }

    pub fn with_persist_path(mut self, path: impl AsRef<Path>) -> Self {
        self.persist_path = Some(path.as_ref().to_path_buf());
        self
    }

    pub fn load_persisted(&mut self) {
        if let Some(path) = &self.persist_path {
            if let Ok(text) = std::fs::read_to_string(path) {
                if let Ok(state) = serde_json::from_str::<PersistedThemeState>(&text) {
                    if self.registry.has(&state.active) {
                        self.active = state.active;
                    }
                    if let Some(mode) = state.lock {
                        self.lock = Some(mode);
                        self.mode = mode;
                    } else if let Some(mode) = state.mode {
                        self.mode = mode;
                    }
                }
            }
        }
    }

    pub fn get_active(&self) -> &str {
        &self.active
    }

    pub fn set(&mut self, id: &str) {
        if !self.registry.has(id) {
            return;
        }
        if self.active == id {
            return;
        }
        self.active = id.to_string();
        self.invalidate_cache();
        self.notify();
        self.persist();
    }

    pub fn get_mode(&self) -> ThemeMode {
        self.mode
    }

    pub fn set_mode(&mut self, mode: ThemeMode) {
        if self.mode == mode {
            return;
        }
        self.mode = mode;
        self.invalidate_cache();
        self.notify();
        if self.lock.is_some() {
            self.persist();
        }
    }

    pub fn get_lock(&self) -> Option<ThemeMode> {
        self.lock
    }

    pub fn lock_mode(&mut self, mode: ThemeMode) {
        self.lock = Some(mode);
        self.mode = mode;
        self.invalidate_cache();
        self.notify();
        self.persist();
    }

    pub fn unlock(&mut self) {
        if self.lock.is_none() {
            return;
        }
        self.lock = None;
        self.notify();
        self.persist();
    }

    pub fn get_all(&self) -> &HashMap<String, ThemeJson> {
        self.registry.get_all()
    }

    pub fn get_theme(&self, id: &str) -> Option<&ThemeJson> {
        self.registry.get(id)
    }

    pub fn get_resolved(&mut self) -> Option<&ResolvedTheme> {
        let cache_key = (self.active.clone(), self.mode);
        if !self.resolved_cache.contains_key(&cache_key) {
            let theme = self.registry.get(&self.active)?;
            let resolved = resolve_theme_safe(theme, self.mode)?;
            self.resolved_cache.insert(cache_key.clone(), resolved);
        }
        self.resolved_cache.get(&cache_key)
    }

    pub fn subscribe(&self) -> broadcast::Receiver<()> {
        self.notify_tx.subscribe()
    }

    fn invalidate_cache(&mut self) {
        self.resolved_cache.clear();
    }

    fn notify(&self) {
        let _ = self.notify_tx.send(());
    }

    fn persist(&self) {
        if let Some(path) = &self.persist_path {
            let state = PersistedThemeState {
                active: self.active.clone(),
                mode: if self.lock.is_some() {
                    Some(self.mode)
                } else {
                    None
                },
                lock: self.lock,
            };
            if let Ok(json) = serde_json::to_string_pretty(&state) {
                let _ = std::fs::create_dir_all(path.parent().unwrap_or(path));
                let _ = std::fs::write(path, json);
            }
        }
    }
}
